### ISSUE TITLE
FolderSelector (Server-side)

### ISSUE THREAD

**Issue by euroblaze** [euroblaze/nextcloud#4]
- [x] Req_1.
Allow option on Chatter to see the server-side folder structure, select folder and upload file directly to it.

- [x] Req_2. 
Search for folder (string search).

**Comment by euroblaze** [comment]
### Actions
- File upload
- Folder upload
- Drag & drop files & folders

[NextCloud-Folder-Uploads.pdf](https://github.com/euroblaze/nextcloud/files/13537721/NextCloud-Folder-Uploads.pdf)

### Storage Struktur
```
- /sales/<id>/..
- /projekt/<id>/.. – /lager/<id>/..
-  ...
```

### Users
- PowerOn System (NC) User 
- Other users can save files and read/write/share
- Share with links

![2023-12-03_01-20-28](https://github.com/euroblaze/nextcloud/assets/7826363/b1415a7a-9165-48a4-bfc5-f74906cc8458)
@holengocquy

**Comment by euroblaze** [comment]
Thanks for the demo on local, @holengocquy 

Next steps:

1. Install on dev/demo or dev2/demo2 by Wednesday (13.12.23).
2. Implement folder uploads (downloads not necessary for now). Use-case: 500 files structured in 10 folders. (Target date: Monday 18.12.23).

Thank you!

**Comment by holengocquy** [comment]
@euroblaze I just deployed to main2 the folder selector feature which I presented to you last week.

**Comment by euroblaze** [comment]
@holengocquy 

Looks like we should take the time & effort to set up demo2.

Reminder: 
main2 is the master copy of the standard software that we deploy for new customers.
DMS/Nextcloud is a PowerPack, sold separately.

Todo:

- [x] setup demo2
- [x] remove DMS/Nextcloud from main2.

Thank you :)

**Comment by holengocquy** [comment]
@euroblaze Ok but where would you like to install demo2?
Is it `116.203.45.134`?

**Comment by euroblaze** [comment]
> Ok but where would you like to install demo2?
> Is it `116.203.45.134`?

How about same machine as demo.pwo?

**Comment by holengocquy** [comment]
Yeah agreed, I mentioned also the ip of `demo.pwo`. Just want to confirm before setting up things.

**Comment by euroblaze** [comment]
On Tue, 12 Dec 2023 at 09:38, 'holengocquy' via euroblaze Github Common
Account ***@***.***> wrote:

> Yeah agreed, I mentioned also the ip of demo.pwo. Just want to confirm
> before setting up things.
>

Go for it..

**Comment by holengocquy** [comment]
> @holengocquy
> 
> Looks like we should take the time & effort to set up demo2.
> 
> Reminder: main2 is the master copy of the standard software that we deploy for new customers. DMS/Nextcloud is a PowerPack, sold separately.
> 
> Todo:
> 
> * [x]  setup demo2
> * [x]  remove DMS/Nextcloud from main2.
> 
> Thank you :)

Done re-setting up demo.pwo. Created new db and installed nextcloud.
There're other modules that may need to be installed, please have a look on `demo.pwo`

**Comment by holengocquy** [comment]
@euroblaze  New db `poweron_demo` created on `demo.pwo`
`admin` / `admin`

**Comment by euroblaze** [comment]
> @euroblaze  New db `poweron_demo` created on `demo.pwo`
> 
> `admin` / `admin`

Login successful. 
Awaiting look & feel and functionality to match main2.

**Comment by euroblaze** [comment]
- [x] Install `powered_by`

@holengocquy

**Comment by euroblaze** [comment]
@holengocquy 

> * File upload
> * Folder upload
> * Drag & drop files & folders

Seeing this UI on demo.
Needs an update?

Thank you for a fix 😄 

![image](https://github.com/euroblaze/nextcloud/assets/7826363/e5dc0f25-0690-4205-91bd-9db3eb11fa26)

**Comment by holengocquy** [comment]
@euroblaze  I resolved the conflict on demo. Module upgraded. Can you try again? Thanks

**Comment by euroblaze** [comment]
On Sun, 17 Dec 2023 at 13:21, 'holengocquy' via euroblaze Github Common
Account ***@***.***> wrote:

> @euroblaze <https://github.com/euroblaze> I resolved the conflict on
> demo. Module upgraded. Can you try again? Thanks
>

It's better.
But still seeing a cached copy of the directory listing..

![image](https://github.com/euroblaze/nextcloud/assets/7826363/a718c977-8fd4-4721-961a-cab2e3ac1b1d)

And if the upload is still attempted, folln. Odoo error throws.

UncaughtPromiseError > ReferenceError
Uncaught Promise > attachmentData is not defined
ReferenceError: attachmentData is not defined
    at FileExploreModel.onUploadClicked (
https://demo.poweron.software/web/assets/570-74a5b84/web.assets_backend.min.js:10506:328
)

**Comment by holengocquy** [comment]
@euroblaze  The issue happens due to this [usecase](https://github.com/euroblaze/nextcloud/issues/8). It's not yet handled. I'll work on it.
In the meanwhile, can you test by using account Qui Ho (I configured back again on demo)

**Comment by euroblaze** [comment]
Getting exhausted testing. Maybe I test 1 time when everything is finished?
Would that work for you also?

**Comment by holengocquy** [comment]
We still have things below to enhance (for huge number of attachments). I was awaiting a greenlight from you.
1. Clean filestore after x days
2. When downloading/uploading a folder, download/upload all its attachments inside
3. [Usecase](https://github.com/euroblaze/nextcloud/issues/8) above

**Comment by euroblaze** [comment]
Thanks for the feedback.

> Clean filestore after x days

Let's not ship this feature in the MVP. Let's postpone to next release, ca. Feb/March 2024.

> When downloading/uploading a folder, download/upload all its attachments inside

Remember having approved this feature. 100% must be done!

> https://github.com/euroblaze/nextcloud/issues/8 above

Yes, this problem must be fixed.
![2023-12-18_16-43-16](https://github.com/euroblaze/nextcloud/assets/7826363/08cc4419-9905-4bd6-8cb6-02c464948b1a).

**Comment by euroblaze** [comment]
@holengocquy 

We decided to descope/suspend **Folder Upload** feature as of today. 

We will return to this feature in the next version of the module, possibly in January 2024. 

Let's ship the rest of the functionality in `demo` before 25.12.23.

**Comment by holengocquy** [comment]
@euroblaze I've just deployed demo an improved version. Could you please take a test? Thanks

**Comment by euroblaze** [comment]
Is there a limit to file-sizes?
Getting this error during upload..

```
UncaughtPromiseError > ReferenceError
Uncaught Promise > attachmentData is not defined
ReferenceError: attachmentData is not defined
    at FileExploreModel.onUploadClicked (https://demo.poweron.software/web/assets/596-3ffc905/web.assets_backend.min.js:10656:328)
```

**Comment by holengocquy** [comment]
Please provide me the file and where did you upload (record model)

**Comment by barrytruong** [comment]
## Step 1
Allow option on Chatter to see the server-side folder structure, select folder and upload file directly to it.

### 1. In model ir.attachment:

Create new fields:
- `parent_id`: many2one field(ir.attachment).
- `child_ids`: One2many(ir.attachment).
–– 
- Add 1 more selection for field type of attachment: [('folder', 'Folder')]
- Add  a new controller to make sure avoiding recursive issue.

New field `parent_id`/`child_ids` will be hidden if `type != folder`

Estimation **4 hours (April 30th - May 1st)**

### 2. In JS part

- for the part of handling of **Chat Post Box**, we need to handle to make the Odoo system understand of "what is folder (`.isDirectory()`)
- Check if attachment is folder then in the route (`/mail/attachment/upload`) controller:
(**Note: also make sure to handle subfolder which are inside of root folder**)

Check if `attachment == True`:
- Create a new ir.attachment with `type == folder` (res_id=??, res_model=??).
- **Base on the model structure now**:  
          ```if attachment.isFile == True, then parent_id = folder attachment
          parent attachment(folder) can contain many attachments(also that is the folder).
          If attachment has no parent, attachment is just a simple file (pdf, png, ... and so on)```

Estimation: **15 hours (May 2nd - 10th).**          
 
## Step 2:

- Handle to download of folder and attachment inside of root folder:
- Add an event download on icon of folder when user click on that, the whole folder will be downloaded.

Estimation: **14 hours (May 12th - 14th)**

## Step 3:

- Handle to click on button to push folder to NextCloud

Estimate **1 hour (May 15th - May 16th).**

## Step 4 (Optional): 

(just a nice to have feature)

- Create a new cron to push to NextCloud automatically everyday (at 12am):
- Create a new field `is_uploaded_to_nextcloud` in `ir.attachment`, 
    so: The cron will check if attachment is not uploaded yet, then upload and set field `is_uploaded_to_nextcloud  = True`.

(**3 hours**)

**Comment by barrytruong** [comment]
@euroblaze I shared with you my idea.

**Comment by euroblaze** [comment]
@Spider4eyes 

The estimation is on the high side.
I can budget max 20h for delivery by 10.05.24.
If agreeable, you can start immediately.

Regards, Ashant
